### PR TITLE
API v1 without api key restriction

### PR DESCRIPTION
**Pull request by fdocr**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature

## Description

After reflecting and discussing internally we're rolling back the decision to restrict CORS + unauthenticated endpoints from the v1 API.

This PR tackles the rollback of that logic without removing the feature flag. Meanwhile #18213 will only include the changes to remove the feature flag (separate PR).

## Related Tickets & Documents

- Related #18213

## QA Instructions, Screenshots, Recordings

1. Enable `:api_v1` Feature Flag
1. Create an API Key in /settings/extensions
1. Send in a PUT request to `/api/users/:id/unpublish` with the API Key and `application/vnd.forem.api-v1+json` in the Accept header
1. Make API requests to the local server
   - The API is exactly the same as v0 (specs are unchanged there) but it should include a few more endpoints regarding unpublishing posts + suspending a user

### UI accessibility concerns?

N/A

## Added/updated tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

- [x] I will share this change internally with the appropriate teams

### CODE REVIEW

**Comment by jaw6** [comment]
I find I'm very confused by our API authentication. Probably I am just missing something, but I'm not sure. 

It seems like *all* of our controllers (being inherited from `ApplicationController`) ought to `verify_private_forem`, which `if api_action?` ought to `authenticate!`, which ought to `authenticate_with_api_key_or_current_user` (with some local differences in v0/v1).

But then also individual controllers might have `before_action :authenticate_with_api_key!, only: %i[create update]` -- is authentication happening twice? is it *actually* not required on some actions?

Since we have v0/v1 redefining `authenticate!` for their own needs, I'm not really understanding why `verify_private_forem` allows for both traditional `current_user` & `api_token` authentication. I would have expected the API parent classes would just redefine this as well? Does this mean that there controllers outside of the `api/v0` or `api/v1` hierarchy that accept token auth?

**Comment by fdocr** [comment]
Hey @jaw6, yeah our authentication logic is somewhat complex but what you're summarizing is pretty accurate.

> It seems like all of our controllers (being inherited from ApplicationController) ought to verify_private_forem, which if api_action? ought to authenticate!, which ought to authenticate_with_api_key_or_current_user (with some local differences in v0/v1).

That's correct. For example DEV isn't a private Forem, so public endpoints are all reachable without authentication (i.e. `GET /api/articles`). This shouldn't happen on private Forems, otherwise the private content would be available via API.

> But then also individual controllers might have before_action :authenticate_with_api_key!, only: %i[create update] -- is authentication happening twice? is it actually not required on some actions?

That is correct, authentication is being checked twice on private Forems. Before this PR we were trying to simplify this on a single call, but we turned that decision around and are coming back to v0 authentication style, and in that process I forgot to include again the `||=` operator we have on v0.

I just added that back again (most recent commit) to avoid having to execute the auth logic if `@user` is already loaded. So thank you for noticing and mentioning this!

In summary again, private Forems will check authentication twice to ensure all endpoints including public ones aren't public (but will now only execute the logic & DB queries once). Non-private Forems like DEV will only check the authentication callbacks on the endpoints that require it on their callbacks.

> I would have expected the API parent classes would just redefine this as well? Does this mean that there controllers outside of the api/v0 or api/v1 hierarchy that accept token auth?

The reason API controllers inherit from `ApplicationController` is somewhat blurry to me as well since it predates my time here as well, but I feel like it works alright in the sense that app controllers (non-API controllers) rely in Devise's `authenticate!` method while API controllers override this to allow for API secret and/or session based auth (@current_user).

But to answer your last question there: No other controllers outside `api/v0` or `api/v1` accept token auth AFAIK.

Let me know if you have any other questions! 🙂

**Review by jaw6 — APPROVED** [review]
:+1: I think I am still finding our authentication methods (especially`verify_private_forem`) hard to parse, but it's totally unrelated to this PR.

**Review by joshpuetz — CHANGES_REQUESTED** [review]
[Request] Can you regenerate the v1 docs (`swagger/v1/api_v1.json`) given they have changed? Also we'll need to make sure they correct denote which API end points require an API key now and which don't.

**Comment by fdocr** [comment]
Great catch @joshpuetz, thanks! I just updated the docs accordingly to show which endpoints require authentication. TIL each unauthorized endpoint can just override the rswag DSL with `security []`

Screen Shot 2022-08-19 at 10 11 51

**Review by joshpuetz — APPROVED** [review]
